Add each_key support to transformers

brutuscat opened the pull request:
Fixes #170

I added support of the key iteration feature into the Proxy and the Transformers, but with caveats because not all the transformers can be "undone" (e.g. not all of the are bijective).

I also had to hack the specs a bit to selectively disable each_key test cases where appropriate, @asppsa please revise and advice.

asppsa commented:
@brutuscat thanks for this, I'll check it out over the weekend!

asppsa commented:
I am inclined towards only supporting `each_key` in those cases where the chain of key transformers can be reversed.  So I think this means all the key transformers need to be either `:encode` or `:compress` types.  If not, then the resulting store should behave as if it doesn't support `each_key`.  

This will work for all the `Moneta.new` cases I can think of off the top of my head, as they mostly just use `Marshal`.

What do you reckon?

brutuscat commented:
@asppsa something like what I just did? I move the compilation into the built transformer class, so it can add the `:each_key` feature at "build time" of the concrete transformer class.

Then it is as you said, either supports it or not based on a rule set in the `config.rb`

brutuscat commented:
@asppsa not sure why mongodb specs are failing...

brutuscat commented:
I reverted my changes to bump the supported ruby version (already some active-support  gems, e.g. rails stuff, are using 2.3) given that for some reason kyotocabinet-ruby-reanimated does not compile in ruby 2.7. 

See https://travis-ci.org/moneta-rb/moneta/jobs/643129451

```
kyotocabinet.cc:4038:7: error: ‘ruby_nonempty_memcpy’ is not a member of ‘std’
745       std::memcpy(kbuf_, kbuf, ksiz);
746       ^
747kyotocabinet.cc:4038:7: note: suggested alternative:
748In file included from
749/home/travis/.rvm/rubies/ruby-2.7.0/include/ruby-2.7.0/ruby.h:33:0,
750                 from kyotocabinet.cc:22:
751/home/travis/.rvm/rubies/ruby-2.7.0/include/ruby-2.7.0/ruby/ruby.h:1758:1: note:
752‘ruby_nonempty_memcpy’
753 ruby_nonempty_memcpy(void *dest, const void *src, size_t n)
754 ^
```

asppsa commented:
@brutuscat I'll open a separate PR for the Ruby version switch-over.  If kyoto cabinet doesn't work any more, we can mark it as unsupported in 2.7+

brutuscat commented:
@asppsa thx. Could you maybe look into what's going on with the MongoDB specs?

asppsa commented:
Sorry it took me so long to check this.  The issue with the MongoDB specs appears to be that they are all using the `"standard_mongo"` database.  When the specs get run in parallel the results will be unpredictable as a result (especially in the case of `each_key`, evidently).  Hopefully this will go away if you make the database name for each file in `spec/moneta/adapters/mongo` different - best practice is to use a database name that corresponds to the spec's filename - e.g. `spec/moneta/adapters/mongo/standard_mongo_moped_spec.rb` should use the `standard_mongo_moped` db.

brutuscat commented:
@asppsa Moped seems to have been forgotten... Since Mongo v3

> Starting in version 3.0, MongoDB deprecates direct access to the system.indexes collection, which had previously been used to list all indexes in a database and in their [code you see](https://github.com/mongoid/moped/blob/master/lib/moped/indexes.rb#L60):

```ruby
    def create(key, options = {})
      spec = options.merge(ns: namespace, key: key)
      spec[:name] ||= key.to_a.join("_")

      database.session.with(write: { w: 1 }) do |_s|
        _s[:"system.indexes"].insert(spec)
      end
    end
```
I will add a warning as it was done before for MongoDB version < 2.2.

brutuscat commented:
@asppsa I need more help with the specs :(

brutuscat commented:
@asppsa I merged master in and fixed the travis issues with couchdb to install and start the db. I also fixed the problems with the Mongo, the Pool and the tcp & unix client specs. 

There was another is tiny fix for Couchdb, but still some specs are failing but does not seem to be my fault... Maybe you could mark these unstable until you figure out a way to make it work with the latest Couchdb? I think v3 has been released and some stuff seemed that changed...

Let me know ;)

brutuscat commented:
@asppsa I think that now I fixed all but the weird problems with mongodb. This group seems to randomly fail only in travis. In my local I see no problem running these... 

Ideas?

brutuscat commented:
@asppsa I fixed the mongo db issues all right! I expect the test suite to be green as of now. Please review

asppsa commented on .rubocop.yml:
```diff
@@ -1,3 +1,4 @@
+require: rubocop-performance
```
Can you remove this for now, and maybe put it into a separate PR?  I need to review what this does.

brutuscat replied:
This are just "cops" for performance. I think that once I upgraded to the latest rubocop,this was once bundled into rubocop and now it is in its own gem and it is optional.

brutuscat replied:
I remember seeing a warning either on the test suite or in my local. But it should cause no harm, given that it is just another part of rubocop.

brutuscat replied:
e.g. https://meta.discourse.org/t/rubocop-performance-gem-warning/116629

asppsa replied:
Ok, that makes sense. :+1:

brutuscat replied:
In the end I think I removed it. I'll make another PR

asppsa commented on lib/moneta/pool.rb:
```diff
@@ -298,6 +298,18 @@ def initialize(options = {}, &block)
+        adapter.each_key(&block)
+        self
+      end
+    end
```
The pool specs were apparently already passing with each key being tested.  What does this add?

brutuscat replied:
I think they were until I fix the proxy for the transformers. The fact that I have to checkout kinda makes sense, because I added that so we have the count optimisation for the enumerator.

I think that if we do not want the "eager loaded" "count" to be passed in, then I could remove this code?

brutuscat replied:
Ok, so if I want to "honor" the "eager load" of the count that the enum_for allows to pass as a block for each adapter that implements it, then when the wrapper (e.g. pool) code runs, that block runs on the scope of the Proxy class and it won't checkout for an adapter.

So I reverted that change in proxy and for now we just keep it like that, is that ok? Or do you prefer me to leave all this code chain here and honor the passed in count for every adapter that implements that?

asppsa commented on lib/moneta/transformer.rb:
```diff
@@ -77,13 +80,26 @@ def without(*options)
+              unless supports? :each_key
+
+            return enum_for(:each_key) { @adapter.each_key.size } unless block_given?
+            @adapter.each_key.lazy.map{ |key| #{key_load} }.each(&block)
```
Would `@adapter.each_key{ |key| yield #{key_load} }` work here?

asppsa replied:
Just thinking about this more:  say for example, you are using a Redis store with the `prefix` key transformer.  If you call `each_key`, it looks like you will get all the keys in the store, whether they are prefixed or not.  I don't think the behaviour of `each_key` in this situation is super-well defined, but I think it's probably at least arguable that a user would expect to only get prefixed keys - what do you think?

brutuscat replied:
I guess you are right, but for now it fits my use case. We could either filter it in weird way inhere or let them know that they have to filter these themselves?

We should probably have a custom block that filters and yields only the ones prefixed?

asppsa commented on .travis.yml:
```diff
@@ -16,17 +16,15 @@ services:
   - memcached
   - mongodb
   - postgresql
+  - couchdb
```
The issue with this approach is it reverts back to CouchDB 1.6, when the current version is now 3.0 ... at least it works though!

I'll see if I can resolve this in another way, but this is a good backup if not.

asppsa commented on spec/support/mongo_helper.rb:
```diff
@@ -0,0 +1,12 @@
+  def self.mongo_config(args)
+    Class.include(MongoHelper).new.mongo_config(args)
+  end
+end
```
To follow the existing convention, this file should be located at `spec/moneta/adapters/mongo/helper.rb`, and required from the specs that need it, instead of from the `spec/helper.rb` file.  Is there a reason you have opted to do it this way?

asppsa replied:
Also, do you have some sense of why we suddenly need this?  Was there a change to how MongoDB works?

asppsa replied:
Actually, I've thought about this some more and I like your approach here better than having the helpers co-located with the specs, so that's a :+1: from me for this change.

asppsa commented on spec/moneta/adapters/client/standard_client_tcp_spec.rb:
```diff
@@ -4,14 +4,14 @@
       Thread.new do
         client = new_store
-        (1..100).each do |j|
+        (1..31).each do |j|
```
I'm not clear on why this change was needed.  Could you give some context?

brutuscat replied:
Not quite sure. I think that in my local (MacOS) there is some kind of limitation with the amount of open ports. 
I figure that if it works with almost 32 it should work with 100...

asppsa commented on spec/moneta/proxies/weak_each_key/weak_each_key_spec.rb:
```diff
@@ -9,8 +9,6 @@
     Moneta.build do
       use :WeakEachKey
-      # use :WeakIncrement
-      # use :WeakCreate
```
I guess these comments were committed by accident? Can we remove them please?

brutuscat replied:
right, I did remove them :)

asppsa commented:
Firstly, a huge thank you for all your work here, and a big apology from me that I've been unable to take a look at this sooner.

What I would really love is if we are able to split out some of the things that are going on in this PR into separate PRs.  In particular the Mongo and Couch stuff should probably be separated, so we can take a look at it in isolation.  There's also a couple of small things like adding `rspec-performance` that I would want to consider more fully before adding.  Let me know if this is too much work though.

The second thing is that I think we need to work out what the expected behaviour of `each_key` is when someone is using different transformers for different keys in the same database.  The obvious example to me is the prefix transformer - one might use the same database with various prefixes in order to separate out the keys.  I guess this is probably applicable to other key transforms as well though.  Possible behaviours that I think would be acceptable are:

* we say that this isn't supported and throw an error if `each_key` encounters a key that it can't transform
*  we support it, and filter out keys that can't be transformed.

Your thoughts on this are appreciated.

Also, I'll see if I can work out a way to get CouchDB 3.0 working ...

asppsa commented:
Actually, the mongo issues are directly connected to the each_key stuff, are they?  In that case I retract my comments about a separate mongo PR!

brutuscat commented:
@asppsa I reverted most of the changes commented here. The good thing is that now we kinda have fixes for all of them ;)

brutuscat commented:
@asppsa alright I fixed the problem of each_key with a "prefix" transformation. I added a "TEST" code in the transformers config and with that it filters the non prefixed keys. Please review.

brutuscat commented:
BTW I tested the couchdb specs in my local with the latest version and it works fine too.

brutuscat commented:
@asppsa regarding the each_key and the transformers:

> we say that this isn't supported and throw an error if each_key encounters a key that it can't transform

I think this only makes sense for prefix. Marshal and other serialisers will fail to "load" and they we will get the exception right there. The same for the encoders I think... 

So I compiled a specific code for these where when these fail if won't break the chain. I also added on marshal->prefix->base64 spec that also tests the each_key with the "wrong" prefix

asppsa commented:
I've fixed CouchDB (and also added Ruby 2.7 testing) in master now.  Hopefully if you merge with master, we'll have this PR passing again.

brutuscat commented on lib/moneta/proxy.rb:
```diff
@@ -22,7 +22,7 @@ def each_key(&block)
       raise NotImplementedError, "each_key is not supported on this proxy" \
         unless supports? :each_key
 
-      return enum_for(:each_key) { adapter.each_key.size } unless block_given?
```
@asppsa so what do you think of this? Is it worth it? Most of the time the adapters will be used via some proxy (e.g. transformers) so I though it would make sense to honor the "eager size" set in the block, this is for the when someone would use the "size" method, see Object#enum_for:
> If a block is given, it will be used to calculate the size of the enumerator without the need to iterate it (see Enumerator#size).

Without this I think size will be always be nil when a proxy is used.

brutuscat commented on lib/moneta/pool.rb:
```diff
@@ -298,18 +298,6 @@ def initialize(options = {}, &block)
-        raise NotImplementedError, "each_key is not supported on this proxy" \
-          unless supports? :each_key
-
-        return enum_for(:each_key) { adapter ? adapter.each_key.size : check_out! { adapter.each_key.size } } unless block_given?
```
@asppsa the change in proxy forces a change here, given that the scope of the block passed in the Proxy class, is within that object, and the checkout never happens.

brutuscat commented:
Retrying the build. There was a weird Ruby 2.3 failure, I think that it was just unfortunate because it passed in all the other builds... The rest seems ok.

asppsa commented:
@brutuscat yes, there is unfortunately some unreliability in the test suite still.  If it happens again this time, I can restart just the test(s) that fail instead of rerunning everything.

Concerning the size stuff, I think it's a good addition.   If the change is required in two files, so be it.

brutuscat commented:
Ok, I added the changes so to honor the adapter eager size block passed in the enum_for. Please review.

brutuscat commented:
@asppsa could u retry the dying Ruby 2.4 specs please? They seemed have timed out or something weird. I see no rspec error trace.

brutuscat commented:
Not at all, LGTM. Thank you!

On Sat, Apr 25, 2020 at 2:02 AM Alastair Pharo <notifications@github.com>
wrote:

> *@asppsa* approved this pull request.
>
> @brutuscat <https://github.com/brutuscat> I am very happy with this now -
> great job! Is there anything else you wanted to add before I merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moneta-rb/moneta/pull/171#pullrequestreview-400330161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACJXBZOB6SZS6YPFQYQ5WTROIR73ANCNFSM4KK2RPEQ>
> .
>


-- 
Mauro Asprea

E-Mail: mauroasprea@gmail.com
Mobile: +34 654297582
Keybase: https://keybase.io/brutuscat
